lib/fs/basicfs_copy_range.go

```diff
@@ -0,0 +1,93 @@
+// Copyright (C) 2019 The Syncthing Authors.
+//
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this file,
+// You can obtain one at https://mozilla.org/MPL/2.0/.
+
+package fs
+
+import (
+	"os"
+	"sort"
+	"strings"
+	"syscall"
+
+	"github.com/syncthing/syncthing/lib/sync"
+	"github.com/syncthing/syncthing/lib/util"
+)
+
+var (
+	copyRangeImplementations []copyRangeImplementation
+	mut                      = sync.NewMutex()
+	implementationOrder      = getCopyOptimisations() // This runs before init
+)
+
+type copyRangeImplementation struct {
+	name string
+	impl func(src, dst basicFile, srcOffset, dstOffset, size int64) error
+}
+
+func registerCopyRangeImplementation(impl copyRangeImplementation) {
+	mut.Lock()
+	defer mut.Unlock()
+
+	if util.StringIndex(implementationOrder, impl.name) == -1 {
+		l.Debugln("Discarding " + impl.name + " copyRange implementation")
+		return
+	}
+
+	l.Debugln("Registering " + impl.name + " copyRange implementation")
+	copyRangeImplementations = append(copyRangeImplementations, impl)
+
+	sort.Slice(copyRangeImplementations, func(i, j int) bool {
+		iidx := util.StringIndex(implementationOrder, copyRangeImplementations[i].name)
+		jidx := util.StringIndex(implementationOrder, copyRangeImplementations[j].name)
+		return iidx < jidx
+	})
+}
+
+// CopyRange tries to use the most optimal way to copy data between two files.
+// Takes size bytes at offset srcOffset from the source file, and copies the data to destination file at offset
+// dstOffset. If required, adjusts the size of the destination file to fit that much data.
+//
+// On Linux/BSD it tries to use ioctl and copy_file_range system calls, which if the underlying filesystem supports it
+// tries referencing existing data in the source file, instead of making a copy and taking up additional space.
+//
+// If that is not possible, the data will be copied using an in-kernel copy (copy_file_range fallback, sendfile),
+// oppose to user space copy, if those system calls are available and supported for the source and target in question.
```

Comment:
Why not fall back to a regular copy in the end? I get that there's no code to use this yet, but we'd have to do the fallback there instead otherwise.

Reply:
I thought this is exactly what we did not want to do, and fail if these fancy things don't work and just write the verified block from ram, oppose to reading it again and writing it back, triggering a race?

Versioner et-al can catch this failure and do the copy themselves, block finder can just write stuff it has in ram.

In practice, I don't think this will ever fail, unless windows/solaris/exotics.